### ISSUE TITLE
Add support for a 'configuration setup'

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1583]
We now have a way to start with any MSX, but it's always empty. On a real MSX setup, you often have several extensions already plugged in, even when the MSX is off. And when you turn it on, the extensions are directly available.

Something similar would be good to have in openMSX as well: have a way to start openMSX with a specific configuration, i.e. a certain MSX with certain extensions inserted. (Technically, one could also include disks inserted, and tapes... as these would also stay in when you power off your real MSX and later turn it back on... but this may be a bit uncomfortable.)

Such a configuration is just a combination of `-machine` and `-extX` command line options. So, just one level higher. Perhaps this information could be stored into a file as a 'setup'? And loaded again?

This would make it much easier to start a specific MSX set up, as you have it at home.... without having to use a workaround (e.g. a savestate or creating your own XML; all having their disadvantages.)

**Comment by MBilderbeek** [comment]
As mentioned in #1768 - ideally the media in the slots should be preserved. Especially for hard drives. (Because in reality, the actual platter with the contents is part of that setup.)

**Comment by balyuf** [comment]
Yes, and perhaps de setup can be described in the main settings file and then selected from the command line or from the menu with just the name of setup.  Such setups could be created by simply saving the current actual setup as a named entry in the settings file ...

**Comment by MBilderbeek** [comment]
The settings.xml would be possible. But also separate files would be possible. Then it is easy to copy them around or share them with others. 
Yes, it should be possible to immediately load them via command line parameters or a console command (and of course also the GUI).

**Comment by MBilderbeek** [comment]
Note that I heard that some people keep using Catapult, just because they can easily start the configuration (selected extensions and machine) from last time, as it is remembered.

@m9710797 I think this is a really clear thing that is currently missing in the new GUI. Do you have some ideas to improve that? Perhaps the stuff from this issue after all?

**Comment by MBilderbeek** [comment]
As a simple alternative: perhaps we can (optionally?) remember the last used configuration (machine + extensions) so that it ca be easily restored?

**Comment by m9710797** [comment]
Before working on an ad-hoc solution for this problem, I think we should first come up with a coherent design for all the session-management related problems. Just a high level design is sufficient (e.g. how will it look/behave from a user point of view). No need to work out any technical or implementation details yet.

Any volunteers?

**Comment by MBilderbeek** [comment]
I did a suggestion in the original post that I think covers all use-cases. It's just about the same combination of machine and extensions that are given a name and can be used again or stored as favorite or pulled from a recent list. (As if you remember the command line that was used to achieve the same.)

I'm very willing/volunteering to discuss these ideas to get them more concrete. It helps if you express your concerns and what you think is still not clear enough.

**Comment by m9710797** [comment]
You're referring to the posts in this thread, right? I only see discussion about the machine configuration.

With _full_ session management I mean including window layout, GUI settings, debugger settings (what's now "imgui.ini" but maybe we should split it over multiple files), the various settings (now in "settings.xml"). Missing parts like breakpoints etc. We may want to include inserted media. Maybe even (optionally) do automatic savestate/loadtstate?

Ideally the user should be able to select which parts of the above should be restored automatically. It may also be good if they can manually be save/restored individually or as group. Maybe we want to be able to combine parts into a new group?

I'm afraid that if we now implement something just for machine configurations. then later it won't fit well with the other parts mentioned above.

It's good to split a large problem into multiple smaller parts. But we should at least have a reasonable idea how (later) all those smaller parts combine into a large whole. (And the pieces should fit reasonably well into that whole).

That what I mean with a coherent full design. Not necessarily work out all the technical details. But we should come up with a reasonable idea for how it may behave, so that the whole makes (fits well together) from a user point of view.

**Comment by MBilderbeek** [comment]
Yes, I wasn't thinking these features were connected.
In fact, I think most of these things seem to be orthogonal to me at first sight. So I didn't connect them. If I'm right, that would help a lot, as it would make it much easier to make smaller steps...

Perhaps we should make an overview what we are talking about in terms of requirements.

Things we'd like to have:
1. be able to start with a certain configuration of an MSX setup, just like on the real thing. It's not only about making it similar to a real situation, it's also relevant if you for instance want to test your own software with the same configuration, over and over again. This is what this GitHub issue is about.
  a. it would be nice to be able to start up with such a setup by default, but this is not a killer
  b. what to do with inserted media? On a real machine, you can keep a floppy in the drive or a cartridge in a slot. In some cases this is useful, but in other cases it is not.
2. be able to save and restore a debug session, so you don't have to remake all your breakpoints, watchpoints, etc. each time you want to start or continue your debugging. This is #1836 . This is about debugger data.
3. be able to save and load different GUI layouts, but without overwriting persistent unrelated data. This is #1698 . E.g. a TAS oriented layout, a debug layout. This is only about ImGUI data.
4. I once created a session management script, but I think no one ever used it. It allows you to continue where you left off using savestates or replays. I guess it's not too useful. 

Anything I missed?

Splitting up imgui.ini seems necessary if only to fix #1698 . And to do serious debugging, you need to be able to save and reload your debugging setup, as per #1836 .

**Comment by m9710797** [comment]
I think (hope) that from an internal implementation point of view most of these thing are indeed orthogonal. But not from a user point of view. Some examples:
* During a debug session (split over two days) I not only want to save my window layout and breakpoints, but I probably also want to continue using the same machine+extension setup.
* Maybe I not only want to split the debug session over multiple days, but interrupt it by playing a game in between.
* Maybe I want different settings for development and playing games. Like scalar=simple, blur=0 for development (to more clearly see pixels), and more realistic looking settings otherwise.
* I may want different debug layouts depending on whether I develop a MSX1 or MSX2 game. (Personally I'd just open the specific windows I need at that time, but I've seen that other people prefer to have a lot of windows open at the same time).

Maybe we need some kind of "profile manager" window in the GUI. It should allow to:
* Create/delete/duplicate profiles, switch between profiles.
* Mix and match partial profiles. E.g. take the window layout from one profile and important it into another.
* Manage the default profile (the one that's used when openMSX starts). Is this a special profile, or just one of the selected profiles?
* What if we exit openMSX, does that update the default profile with the current settings or update the selected profile? Should it be configurable which aspects get automatically saved? Which changes are lost on exit.
* It should be possible for a profile to have a separate settings.xml from the default. IOW it should both be possible to share parts (like settings.xml) between (some) profiles, but also have it specific to a single profile.
* ...
* Ah, and it should still be intuitive for the user to configure :)

In other words: I think we should first get a rough idea about how such a full profile manager could look like and what possibilities it should offer. And only then we can start to implement the smaller parts of it.

**Comment by MBilderbeek** [comment]
I mean, this would be a solution, but aren't there simpler ones?

Like:
* To continue with the same machine+extension setup, it would be fine to be able to quickly select it from a 'recently used' drop down or even start up with a default setup. This is basically what is/was offered by Catapult as well. So that would match the existing functionality of 19.1. The point of course was that people never had to restart Catapult on the same day, so they kept things selected as they were[1]. Note that this includes the selected media! (So, include that in the setup after all??)
* People were used to load debugger sessions in the old debugger, to get their loaded symbols and breakpoints etc. back. If you can just easily reload your session, you can continue quickly.
* AFAIK the layout wasn't included in the debugger session and is just fixed/remembered - but OTOH, you almost never had to restart the debugger over openMSX sessions [2]) either, so you just kept it as it was and reconnected to openMSX. So maybe it's fine to be able to get the last used layout restored and then be able to quickly change it.

Regarding [1] and [2]: one of the biggest differences with the openMSX 19.1 situation is that the old separate programs were kept running and remembered some settings for the user. So, for Catapult: which media to use, which extensions to use, which machine to use, and with that the user got it back by just pressing the Catapult Start button. And for the debugger similarly.
One of the reasons that (especially on Windows) people have to restart openMSX often is when developing a ROM game due to #955 . With the mechanism I just described this can be more or less worked around, but now that it's all integrated, you lose a lot of state when starting up openMSX each time.

With these points I'm *only* talking about functional regressions in openMSX 20.0 compared to 19.1 + Catapult + Debugger. I think it's most important to cover these first and not think about new requirements (that I have never heard anyone ask for so far) like different scaler settings depending on what you are doing.

**Comment by balyuf** [comment]
I am all for the idea of first trying to solve the functional regressions indeed.  Recently used, is good enough.  Perhaps add an option to always start openMSX with a dialogue to select a recently used setup before starting up the machine?  In that dialogue the default selection could be the last used setup.  This would mimic closely the functionality we had in Catapult.

**Comment by MBilderbeek** [comment]
> Perhaps add an option to always start openMSX with a dialogue to select a recently used setup before starting up the machine? In that dialogue the default selection could be the last used setup. This would mimic closely the functionality we had in Catapult.

I'm not sure that's a good idea. In my most common use case that would always block me from going forward, so it would be rather annoying. But feel free to come up with more ideas :)

**Comment by balyuf** [comment]
> > Perhaps add an option to always start openMSX with a dialogue to select a recently used setup before starting up the machine? In that dialogue the default selection could be the last used setup. This would mimic closely the functionality we had in Catapult.
> 
> I'm not sure that's a good idea. In my most common use case that would always block me from going forward, so it would be rather annoying. But feel free to come up with more ideas :)

Well, that's why I formulated as "add an option", for those that do find it useful.  For the others the recent list would suffice?

**Comment by MBilderbeek** [comment]
Well, as I said, let's first focus on actual functional regressions people are running into. @m9710797 can you comment on my post of the 23rd of February?

**Comment by MBilderbeek** [comment]
@m9710797 pretty please?

**Comment by m9710797** [comment]
@MBilderbeek if someone (you?) makes a prototype or a reasonably detailed proposal, then I can help implement it. But this is not something I'll start working on myself (or at least not soon).

**Comment by MBilderbeek** [comment]
> [@MBilderbeek](https://github.com/MBilderbeek) if someone (you?) makes a prototype or a reasonably detailed proposal, then I can help implement it. But this is not something I'll start working on myself (or at least not soon).

OK, so phase 1, the original proposal:
- we add a "setup" concept. To start simple, a "setup" is simply the current machine and inserted extensions
- we store the machine and insertions in a simple text file in a specific directory `setups` (alongside `savestates`), could be simply commands:
  - `machine Toshiba_HX-10`
  - `exta fmpac`
- we add commands to load setups and save the current setup
- we add menu options to load setups and save the current setup, including recently used setups.
- setups can have custom names (filename), which could be suggested by openMSX based on the current machine name and inserted extension(s)

This is not full session management, and I don't think it should be. This is solving a real problem that users have over and over again. Very recent example: https://www.msx.org/forum/msx-talk/openmsx/how-to-bind-quick-loadsave-state-to-a-button-and-other-questions

We could extend it to also preserve the content of media, by adding storing of the commands for that as well:
- `hda myharddisk.dsk`
- `diska mydisk.dmk`
(I'm not sure if we can distinguish extensions from inserted ROM cartridges, by the way. But then we just get these as well.)

And perhaps also add the plugged hardware:
- `plug joyporta msxjoystick1`
- `plug joyportb mouse`

@m9710797 is this concrete enough?

**Comment by m9710797** [comment]
My concern with such ad-hoc implementations (that is solve the immediate problem, without thinking about the future full solution) is that it we should keep supporting it "forever" and that we have no idea if it will fit well with the future full solution. But OK, I'm not coming with ideas for a full solution. And at least you're thinking about this partial solution. So let's go with that.

So a crude summary of this idea is to represent the setup as a list of Tcl commands (a script?) that reconstruct the desired state? That could indeed be extended in the future with more commands.

I was thinking: we could in the future even include a savestate. But soon after I realized that a savestate makes most (all?) other commands in the script obsolete. And then maybe we could instead of a script just always use the existing savestate infrastructure? If you don't want the 'savestate'-part, just the 'setup'-part, then you can do a "power-off / power-on" to obtain that. If you don't want to include inserted media, you can eject that media right before creating the setup-file.

I've only just came up with this idea (I've only thought about this for a few minutes), maybe there is some flaw in this idea. But the advantage of using savestates, is that it's existing infrastructure, and it (should) already support everything. And savestate do have support for backwards compatibility. And we don't invent some totally new concept (that we need to keep supporting "forever").

**Comment by MBilderbeek** [comment]
Using a savestate is actually currently one of the workarounds I mentioned in the reply to J-War's post on MRC which I referred to in my previous post. And yes, I think it might just do the same thing, in the end.

However, although we would use exactly the same mechanism as a savestate, we should not present it as a savestate. In the end, the mechanism is just the means to make the 'setup' concept. 

So, can we then express the proposal as follows:
- we add a "setup" concept. To start simple, a "setup" is simply the current machine and inserted extensions
- internally, a setup is simply just a savestate directly after power on, with all media ejected (we might want to make this optional) stored in a specific directory `setups` (alongside `savestates`)
- we add commands to load setups and save the current setup
- we add menu options to load setups and save the current setup, including recently used setups.
- setups can have custom names (filename), which could be suggested by openMSX based on the current machine name and inserted extension(s)

Like that?

**Comment by m9710797** [comment]
Yes. Setups are a new user concept, but internally (as a first attempt) we'll try to implement them as savestates.

Maybe for now we can leave out your 3rd point (create commands for load/save setups) (we can instead issue savestate commands)? At least for the prototype, if it works out fine, we can still decide whether or now we want dedicated commands.

**Comment by MBilderbeek** [comment]
Oh, one drawback of using savestates internally: if we later update a machine configuration or any of the used extension configurations, they will not be used automatically in the setup, as it's part of the savestate.
What do you think, is that a reason not to go this way?

**Comment by MBilderbeek** [comment]
As discussed offline:
- if we construct the savestates by reconstructing it (load the same machine (which will get the latest config), then load the plugged in extensions (which will also then get the latest config) and pluggables and media, we seem to be fine, i.e. we can implement what we want using the savestates system without losing updates to configurations as mentioned in my last comment
- when saving a setup, we could optionally offer different hierarchical layers of saving (in order of dependency - you can't skip layers):
  - machine (obviously)
  - extensions (typically, otherwise there's no added value)
  - pluggables (typically, but maybe some would like to leave them out)
  - media (useful depending on the use case - probably disable by default)
-  it seems to be good to be able to start up automatically with the last used setup (as such it should be saved at exit)
  - if such a setup is found, it overrides the `default_machine` setting
  - if this is disabled, and a `default_setup` setting has a value, we should start up with that setup instead of the `default_machine`
- we need to add a `-setup` command line option that allows you to start with a certain setup (overriding any `default_setup` or `default_machine` setting)
  - it is an error if both a `-setup` and `-machine` option are given
  - if also `-ext` or `-diska` or other extension/media options are given, these will override stuff from the given `-setup` option.
- in the GUI, we can offer menu options like this:
  - machine -> Load setup...
    - when loading a setup, you can optionally skip any of the layers (respecting the hierarchy), by default all are loaded. This is shown in a preview of the setup to be loaded.
    - a list of recently used setups must be shown
  - machine -> Save current setup...
    - the name is freely choosable, we could pre-fill in a suggestion based on what's in the machine (machine name + extensions?)
    - the hierarchy of layers must be shown to select what to include in the setup. You can (de)select whole layers or individual items in them. Deselecting a higher level layer automatically disables selecting any lower layer)
      - machine, including the description as displayed in the machine list
      - extension list (with slots, including description as displayed in the extension selector)
      - list of plugged in pluggables 
  - machine -> Save setup as... (perhaps we can somehow combine it with the previous)
  - also we must offer somewhere what must be done at startup:
    - load last setup
    - load default setup (initially, the default setup could be just an empty instance of the default machine). Note that the default machine stuff can be removed from the Select MSX Machine dialog, as it is then replace with default setup.
    - or even optionally: open load setup dialog
  - and it must be configurable which layers are saved automatically when openMSX is exited in the 'last setup'.

**Comment by m9710797** [comment]
I mostly agree with MBilderbeek's summary. Here are some small additions/clarifications. And some are actually more questions than fully worked out ideas.
* (From a user point of view, not implementation) I see a 'setup' as an extended version of a 'machine'. A 'machine' is a special case of a 'setup', one with only a machine, but no extensions, pluggables or media. So a setup with only the first layer present.
  * What we currently do for the default 'machine' at startup, will be replaced by a 'setup'. So next to a 'default_machine' setting, we'll also have a 'default_setup' setting. If both are present, then 'default_setup' takes priority.
  * On the command line next to the '-machine' option, there will also be a '-setup' option. Either of these options overrides the 'default_machine' or 'default_setup' setting. But you cannot use both command line options simultaneous.
* About 'startup with the last used machine'. Technically that would be equivalent to automatically updating the value of the 'default_setup' setting (when switching to a new setup). We currently don't have an equivalent function for the default_machine (and we shouldn't add it for machine) But I do think that's useful new functionality.
  * This could be enabled/disabled via a new boolean setting. Or it could be some convention on the name (e.g. "my-setup" means a fixed setup, but "*my-setup" means it's automatically updated). It doesn't really matter how it's mapped on setting(s), logically it's a boolean and a string.
  * How does this interact with switching to a new machine? If we view machines as a simple version of a setup (one with just the first layer), then switching to a new machine should also update the last used setup.
* Hmm, if we accept this last point, then maybe, from a user point of view, it's simpler to not distinguish between machines and setups at all. So unify the 'default_machine' and 'default_setup' into a single setting (and for backwards-compatibility keep the name 'default_machine'). That also eliminates the complication that "setup" has higher priority than "machine". Same for the '-machine' and '-setup' command line options. From an implementation point of view, we should then of course lookup the name both as a (pure) machine or as a setup.
* Next to remembering the last used setup-name there's an orthogonal concept of saving (=creating) a setup when you exit openMSX. So when I insert new extensions/pluggables/media during this session, next time I start openMSX it restores those changes. I see two ways to offer this functionality:
   * Replace the above boolean with an enum with 3 possible values:
      * Don't remember the last used setup.
      * Remember the last used setup (only the name of that setup, don't update that setup).
      * Remember and also update the last used setup.
   * Keep the boolean as above, but now add a new string-setting: the name of the setup file to write on exit.
      * E.g. if you set both this new 'save_setup' and the 'default_setup' (or default_machine?) settings to the same string (e.g. "my-last-session", then you'll get the behavior of an automatically updating setup.
      * This is the most flexible approach, but maybe also the most difficult to understand for the user?
* UPDATE: when I re-read the above items, I now think that (only) updating the last used setup-name isn't need anymore once you have the functionality of create a whole new setup file on exit. But I kept the above points anyway (if only to make the distinction clear).
* About the hierarchy:
  * MBilderbeek's summary mentioned:
    * When explaining the layers: the hierarchy is strict (you can't skip layers).
    * On loading you can skip layers (all lower layers, so still respecting the hierarchy).
    * On creating a new setup you can deselect individual items (skipping levels in the hierarchy).
  * I don't agree with the last item "deselecting items that skip the hierarchy". It's too easy to create mistakes like this. E.g. deselect a FDD extension, but keeping the media for that disk drive (and not easy for us to detect/flag such mistakes).
  * I'm also hesitant to allow skipping layers on load. Because that feature requires to "undo" or "erase" properties from a savestate. E.g. technically we could load a savestate and then eject the media (or unplug extensions etc). But that first step may already fail because the media can not be found. From a user point of view that's a stupid error because he has chosen to skip that media. It's not so easy to already skip during loading. So for sure in a first version I'd leave this out.
* Some open questions:
  * Should we unify 'machine' and 'setup'?
    *  If you don't take saving into account, then (to me) this seems like a nice solution.
    *  But if you can also save the (modified) setup on exit, then this poses questions of what name to use. We can't override the existing machine name. So should we invent a new name?
  * (Whether or not we unify machine and setup). Suppose you start from an existing machine or setup. And you have create-setup-on-exit enabled. Will that overwrite the setup you started from, or will it create a new one?
    * And if you create a new one, how do we avoid an ever growing number of setups?
    * Maybe there should be a single "last-setup". I mean with a fixed name (actually that comes close to my proposal from above for a "save_setup" setting, so maybe that shouldn't be a setting at all?)

**Comment by MBilderbeek** [comment]
> I mostly agree with MBilderbeek's summary. Here are some small additions/clarifications. And some are actually more questions than fully worked out ideas.
> 
>     * (From a user point of view, not implementation) I see a 'setup' as an extended version of a 'machine'. A 'machine' is a special case of a 'setup', one with only a machine, but no extensions, pluggables or media. So a setup with only the first layer present.
>       
>       * What we currently do for the default 'machine' at startup, will be replaced by a 'setup'. So next to a 'default_machine' setting, we'll also have a 'default_setup' setting. If both are present, then 'default_setup' takes priority.
>       * On the command line next to the '-machine' option, there will also be a '-setup' option. Either of these options overrides the 'default_machine' or 'default_setup' setting. But you cannot use both command line options simultaneous.

This is exactly equivalent to what I put in my text above, right? 

>     * About 'startup with the last used machine'. Technically that would be equivalent to automatically updating the value of the 'default_setup' setting (when switching to a new setup). We currently don't have an equivalent function for the default_machine (and we shouldn't add it for machine) But I do think that's useful new functionality.
>       
>       * This could be enabled/disabled via a new boolean setting. Or it could be some convention on the name (e.g. "my-setup" means a fixed setup, but "*my-setup" means it's automatically updated). It doesn't really matter how it's mapped on setting(s), logically it's a boolean and a string.
>       * How does this interact with switching to a new machine? If we view machines as a simple version of a setup (one with just the first layer), then switching to a new machine should also update the last used setup.
> 
>     * Hmm, if we accept this last point, then maybe, from a user point of view, it's simpler to not distinguish between machines and setups at all. So unify the 'default_machine' and 'default_setup' into a single setting (and for backwards-compatibility keep the name 'default_machine'). That also eliminates the complication that "setup" has higher priority than "machine". Same for the '-machine' and '-setup' command line options. From an implementation point of view, we should then of course lookup the name both as a (pure) machine or as a setup.
> 
>     * Next to remembering the last used setup-name there's an orthogonal concept of saving (=creating) a setup when you exit openMSX. So when I insert new extensions/pluggables/media during this session, next time I start openMSX it restores those changes. I see two ways to offer this functionality:
>       
>       * Replace the above boolean with an enum with 3 possible values:
>         
>         * Don't remember the last used setup.
>         * Remember the last used setup (only the name of that setup, don't update that setup).
>         * Remember and also update the last used setup.
>       * Keep the boolean as above, but now add a new string-setting: the name of the setup file to write on exit.
>         
>         * E.g. if you set both this new 'save_setup' and the 'default_setup' (or default_machine?) settings to the same string (e.g. "my-last-session", then you'll get the behavior of an automatically updating setup.
>         * This is the most flexible approach, but maybe also the most difficult to understand for the user?
> 
>     * UPDATE: when I re-read the above items, I now think that (only) updating the last used setup-name isn't need anymore once you have the functionality of create a whole new setup file on exit. But I kept the above points anyway (if only to make the distinction clear).
> 
>     * About the hierarchy:
>       
>       * MBilderbeek's summary mentioned:
>         
>         * When explaining the layers: the hierarchy is strict (you can't skip layers).
>         * On loading you can skip layers (all lower layers, so still respecting the hierarchy).
>         * On creating a new setup you can deselect individual items (skipping levels in the hierarchy).
>       * I don't agree with the last item "deselecting items that skip the hierarchy". It's too easy to create mistakes like this. E.g. deselect a FDD extension, but keeping the media for that disk drive (and not easy for us to detect/flag such mistakes).

I'm not sure what you mean here. I never tried to suggest that the hierarchy can be skipped. I mean that if you don't want to restore all items from the *leaf* layer, that should be possible. I hope that clarifies it.

>       * I'm also hesitant to allow skipping layers on load. Because that feature requires to "undo" or "erase" properties from a savestate. E.g. technically we could load a savestate and then eject the media (or unplug extensions etc). But that first step may already fail because the media can not be found. From a user point of view that's a stupid error because he has chosen to skip that media. It's not so easy to already skip during loading. So for sure in a first version I'd leave this out.

Fine to leave this out in the first version. We'll see later how it goes. The user can always manually remove extensions after loading the setup.

>     * Some open questions:
>       
>       * Should we unify 'machine' and 'setup'?
>         
>         * If you don't take saving into account, then (to me) this seems like a nice solution.

One question: if there's a setup with the same name as a machine, and the setup is more than just that machine, what do you expect to happen?
Saving setups could 'hide' bare machine configurations. So probably we should always show both in a selection dialog and the preview (showing the properties) should make clear what is the difference.

Also, from a technical point of view, the machines will remain as they are, i.e. the definitions, file format and location. But they can also be loaded as setups that are just layer 1 (a bare machine).

>         * But if you can also save the (modified) setup on exit, then this poses questions of what name to use. We can't override the existing machine name. So should we invent a new name?
>       * (Whether or not we unify machine and setup). Suppose you start from an existing machine or setup. And you have create-setup-on-exit enabled. Will that overwrite the setup you started from, or will it create a new one?
>         
>         * And if you create a new one, how do we avoid an ever growing number of setups?
>         * Maybe there should be a single "last-setup". I mean with a fixed name (actually that comes close to my proposal from above for a "save_setup" setting, so maybe that shouldn't be a setting at all?)

Like `quicksave`: I had in mind that the last-setup is saved under a fixed name only to restore the setup that was last used. Then it's transparent. Named setups are setups named by users themselves. openMSX only helps suggesting a useful name.

One extra concern: if you have multiple machines running, what should we do with this feature? Have multiple 'last-used'? I guess so. Note that this functionality isn't very commonly used, I think, but still we should consider it and handle it properly.

**Comment by MBilderbeek** [comment]
TODO: 
- [x] use the setup overview as preview when loading setups: process review comments.
- [x] add selection for when saving setups to specify the level to save
- [x] fix the display of the machine config details in the Load Setup preview, when expanding the 'Machine' node
- [ ] take care of the windows/menus not getting too wide. 
- [ ] it seems that sometimes the items in the Load Setup list do not respond when being clicked on, i.e. nothing happens instead of showing the preview... find out what is going wrong.
- [x] fix: when I started up with a Boosted MSX2 EN and click on a setup (in Load Setups) where an LD is inserted and then expand the Machine treenode, the automatic laserdisc loading instruction gets typed in the currently running machine: I see `1CALLLD` being entered in the machine....
- [x] make it possible to make a certain setup the default one (edit the `default_setup` setting in the GUI).
- [ ] ~~perhaps rename "Load setup..." to "Setups...", showing the saved setups and a way to load them or make them default~~.
- [x] beautify the overview by using tables like the pluggables
- [x] take a look at when we use " ..." in menu items. It seems inconsistent
- [ ] possibly move setup stuff from menu item to window. This will make sense if we would also allow to do (small) edits to the setup
- [x] when saving setup, use the human readable suggested name instead of the config name
- [x] add the functionality to optionally save the current setup at exit using a (fixed?) filename. The setting contains the level to save, including NONE to disable it.
- [x] ~~For the GUI of auto-save: one idea is to be able to mark the current setup being saved as a setup that should always be saved under that name at exit. Under water, the selected level will then also be used for the auto-save. How to disable this auto-save: Maybe (when auto-save is enabled) we can add an extra status line in the "Save Setup" menu that shows the current save-name (and maybe also level?).  Maybe it's sufficient to add a button "Disable" there?   Or optionally you could allow to edit the properties there (instead of the disable button)~~ Replaced by Setup settings menu item.
